packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+        destructure.type ??= getConstrainedTypeAtLocation(
+          services,
+          destructure.property.key,
+        );
+
+        for (const type of tsutils.unionTypeParts(destructure.type)) {
+          if (tsutils.isStringLiteralType(type) && type.value === memberKey) {
+            return destructure.property;
+          }
+
+          if (tsutils.isNumberLiteralType(type) && type.value === memberKey) {
+            return destructure.property;
+          }
+        }
+      }
+
+      // compare types for computed type keys
+      if (member.computed) {
+        const memberKeyType = services.getTypeAtLocation(member.key);
+
+        for (const destructure of dynamicProperties) {
+          destructure.type ??= getConstrainedTypeAtLocation(
+            services,
+            destructure.property.key,
+          );
+
+          for (const type of tsutils.unionTypeParts(destructure.type)) {
+            if (checker.isTypeAssignableTo(type, memberKeyType)) {
+              return destructure.property;
+            }
+          }
+        }
+      }
+
+      return undefined;
+    }
+
+    /**
+     * @returns `true` if an index signature's key matches an existing
+     * property's key, `false` otherwise
+     */
+    function indexSignatureHasMatchingKey(
+      indexParameterType: ts.Type,
+      dynamicProperties: DynamicProperties,
+      remainingProperties: RemainingProperties,
+    ) {
+      const indexParameterTypeParts =
+        tsutils.unionTypeParts(indexParameterType);
+
+      const indexParameterHasNumberType = indexParameterTypeParts.some(
+        tsutils.isIntrinsicNumberType,
+      );
+
+      const indexParameterHasStringType = indexParameterTypeParts.some(
+        tsutils.isIntrinsicStringType,
+      );
+
+      // check remaining properties for a match
+      for (const key of remainingProperties.keys()) {
+        if (typeof key === 'number' && indexParameterHasNumberType) {
+          return true;
+        }
+
+        if (typeof key === 'string' && indexParameterHasStringType) {
+          return true;
+        }
+      }
+
+      // check dynamic properties for a match
+      for (const destructure of dynamicProperties) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.key,
+        );
+
+        for (const type of tsutils.unionTypeParts(destructure.type)) {
+          if (checker.isTypeAssignableTo(type, indexParameterType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * @returns `true` if an index signature's value type matches an existing
+     * property's value type, `false` otherwise
+     */
+    function indexSignatureHasMatchingValue(
+      remainingProperties: RemainingProperties,
+      indexSignature: TSESTree.TSIndexSignature,
+    ): boolean {
+      // `{ used }: { [i: `_${string}`] }`
+      if (!indexSignature.typeAnnotation) {
+        return true;
```

Comment:
[Testing] Uncovered line, as reported by codecov. Either this is reachable and should have a test covering it, or it's not and either a refactor or type assertion should be used to eliminate the dead code.